lib/idle_task_queue.dart

```diff
@@ -0,0 +1,67 @@
+// @JS()
+library idle_task_queue;
+
+import 'dart:async';
+import 'dart:html';
+import 'dart:js';
+
+import 'package:meta/meta.dart';
+
+/// A handler which will be called during idle time.
+typedef void IdleTaskHandler<T>(T input);
+
+@immutable
+class _IdleTask<T> {
+  final IdleTaskHandler<T> handler;
+  final T input;
+
+  _IdleTask(this.handler, this.input);
+}
+
+/// Adds the task to a queue which will be called during idle time.
+void enqueueIdleTask<T>(IdleTaskHandler<T> handler, T input) {
+  // For unsupported browsers when the polyfill is not available,
+  // we still want the task to be called eventually.
+  if (!_doesBrowserSupportIdleCallback) {
+    new Future.delayed(const Duration()).then((_) => handler(input));
+    return;
+  }
+
+  _taskQueue.add(new _IdleTask<T>(handler, input));
+
+  // Start running the queue if it has not been started.
+  // If there is a task, the queue is already running and doesn't
+  // need to be started again.
+  if (_currentTask == null) {
+    _currentTask = window.requestIdleCallback(
+      allowInterop(_runQueue),
+      <String, dynamic>{'timeout': 1000},
+    );
+  }
+}
+
+void _runQueue(IdleDeadline deadline) {
+  while (!deadline.didTimeout &&
+      deadline.timeRemaining() > 0 &&
```

Comment:
I wonder if we should give a few more ms here. Should we start another callback if we have 1ms left? Seems likely it will overflow the frame. But it obviously depends on how much work the callback is doing. 
Future: Can we send the amount of time remaining to the callback so it can bail early?

Reply:
I've added the future idea to the ticket linked above. As for your first question, a few ms might make sense here. Ideas on a range? I'd hesitate to make this more than a few ms because otherwise we might run into an issue where jobs are unexpectedly not getting done and consumers of the API would need to be made aware that the buffer exists to explain why.

Reply:
Right, we don't want to starve the queue. Maybe just leave it as is for now and see if we actually ever blow out frames because of this. It could not be an issue in practice.